dlfcn.c

```diff
@@ -463,6 +465,92 @@ char *dlerror( void )
     return error_buffer;
 }
 
+#ifdef __USE_GNU
+/*
+ * Return adress from Image Allocation Table (iat), if
+ * the original address points to a thunk table entry.
+ */
+static void *getAddressFromIAT( const void *addr )
+{
+    const unsigned char *p = addr;
+    /* ...inline app code...
+     * 00401002  |. E8 7B0D0000    CALL 00401D82               ; \GetModuleHandleA
+     * ...thunk table...
+     * 00401D82   $-FF25 4C204000 , JMP DWORD PTR DS:[40204C]  ;  KERNEL32.GetModuleHandleA
+     * ...memory address value of pointer...
+     * 40204C > FC 3D 57 7C   ;little endian pointer value
+     */
+    if( *p == 0xff && *(p+1) == 0x25 )
+    {
+        void **p1 = (void *)(p+2);
+        void **p2 = *p1;
+        /* TODO check that *p2 is inside IAT*/
+        /* fprintf(stderr, "%p -> %p\n", addr, *p2); */
+        return *p2;
+    }
+    return NULL;
+}
+
+/* holds module filename */
+static char _module_filename[2*MAX_PATH];
+
+/**
+ * Get module information (filename and base address) from the address given
+ * @param addr address to get module info for
+ * @param info pointer to store module info
+ * @return 0 requested info filled into structure pointed by parameter info
+ * @return 1 error
+ */
+static int getModuleInfo(const void *addr, Dl_info *info )
+{
+    HMODULE hModule;
+    int sLen;
+
+    if (!GetModuleHandleExA( GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, addr, &hModule))
+        return 1;
+
+    if( !hModule )
+    {
+        SetLastError( ERROR_INVALID_PARAMETER );

```

Comment:
`SetLastError` is suspicious in `dladdr` as nobody would expect that POSIX/GNU dladdr error should be returned by `GetLastError`.